sceptre_cdk_handler/cdk_builder.py

```diff
@@ -0,0 +1,214 @@
+import logging
+import os
+import subprocess
+import sys
+from abc import ABC, abstractmethod
+from typing import Any, Optional, Dict, Type
+
+import aws_cdk
+from aws_cdk.cx_api import CloudAssembly
+from botocore.credentials import Credentials
+from cdk_bootstrapless_synthesizer import BootstraplessStackSynthesizer
+from sceptre import exceptions
+from sceptre.connection_manager import ConnectionManager
+from sceptre.exceptions import TemplateHandlerArgumentsInvalidError
+
+
+class SceptreCdkStack(aws_cdk.Stack):
+    def __init__(self, scope: aws_cdk.App, id: str, sceptre_user_data: Any, **kwargs):
+        super().__init__(scope, id, **kwargs)
+        self.sceptre_user_data = sceptre_user_data
+
+
+class CdkBuilder(ABC):
+    STACK_LOGICAL_ID = 'CDKStack'
+
+    @abstractmethod
+    def build_template(
+        self,
+        stack_class: Type[SceptreCdkStack],
+        cdk_context: Optional[dict],
+        sceptre_user_data: Any
+    ) -> dict: ...
+
+
+class BootstrappedCdkBuilder(CdkBuilder):
+
+    def __init__(
+        self,
+        logger: logging.Logger,
+        connection_manager: ConnectionManager,
+        *,
+        subprocess_run=subprocess.run,
+        app_class=aws_cdk.App,
+        environment_variables=os.environ
+    ):
+        self._logger = logger
+        self._connection_manager = connection_manager
+        self._subprocess_run = subprocess_run
+        self._app_class = app_class
+        self._environment_variables = environment_variables
+
+    def build_template(
+        self,
+        stack_class: Type[SceptreCdkStack],
+        cdk_context: Optional[dict],
+        sceptre_user_data: Any
+    ) -> dict:
+        assembly = self._synthesize(stack_class, cdk_context, sceptre_user_data)
+        self._publish(assembly)
```

Comment:
wouldn't this do a publish even if nothing has changed?  is AWS idempotent for this situation?

Reply:
Yes, it would, I think. Unless the `cdk-assets` library is smarter than that. But I think it would always push. SAM CLI does something similar. I _believe_, however, that the filename is based on a hash of the contents, which would mean that for file assets, it would simply re-put the same file at the same key. And for ECR, docker is smart enough to know when image layers have already been pushed and will skip those. So I don't think it's the end of the world to do it this way.

Reply:
sounds good.